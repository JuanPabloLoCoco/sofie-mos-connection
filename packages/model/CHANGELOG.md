# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.7](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.6...v3.0.7) (2023-12-27)

**Note:** Version bump only for package @mos-connection/model





## [3.0.5](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.4...3.0.5) (2023-12-18)


### Bug Fixes

* Handle empty MOS time string ([85fbc88](https://github.com/nrkno/sofie-mos-connection/commit/85fbc886d7b577db07bece23efc53f6058a92a43))





## [3.0.4](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.1...v3.0.4) (2023-06-09)

**Note:** Version bump only for package @mos-connection/model





## [3.0.2](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.1...v3.0.2) (2023-03-27)

**Note:** Version bump only for package @mos-connection/model





# [3.0.0](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.0-alpha.3...v3.0.0) (2023-02-03)

**Note:** Version bump only for package @mos-connection/model





# [3.0.0-alpha.3](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.0-alpha.2...v3.0.0-alpha.3) (2023-01-27)

**Note:** Version bump only for package @mos-connection/model





# [3.0.0-alpha.2](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.0-alpha.0...v3.0.0-alpha.2) (2023-01-27)


### Bug Fixes

* add stringifyMosObject, an utility-function used to convert objects containing IMOSString128 etc to strings ([f3806ab](https://github.com/nrkno/sofie-mos-connection/commit/f3806ab4e72a02b450e91ab19fbbfca34c605caa))





# [3.0.0-alpha.1](https://github.com/nrkno/sofie-mos-connection/compare/v3.0.0-alpha.0...3.0.0-alpha.1) (2023-01-27)


### Bug Fixes

* add stringifyMosObject, an utility-function used to convert objects containing IMOSString128 etc to strings ([f3806ab](https://github.com/nrkno/sofie-mos-connection/commit/f3806ab4e72a02b450e91ab19fbbfca34c605caa))





# [v3.0.0-alpha.0](https://github.com/nrkno/sofie-mos-connection/compare/2.0.1...v3.0.0-alpha.0) (2022-12-09)

### Features

- move helper functions into a separate package: $mos-connection/helper ([ecb51ec](https://github.com/nrkno/sofie-mos-connection/commit/ecb51ec3ca26c15a61fd629e59265345c247f82e))
- move types and enums to @mos-connection/model ([2266488](https://github.com/nrkno/sofie-mos-connection/commit/2266488f4062da6a1f2949a3374c58c26a20d79e))
