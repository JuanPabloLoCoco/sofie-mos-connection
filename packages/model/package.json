{
	"name": "@mos-connection/model",
	"version": "3.0.4",
	"description": "Types and enums for the mos-connection library",
	"main": "dist/index.js",
	"typings": "dist/index.d.ts",
	"license": "MIT",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/nrkno/sofie-mos-connection.git"
	},
	"homepage": "https://github.com/nrkno/sofie-mos-connection#readme",
	"scripts": {
		"build": "rimraf dist && yarn build:main",
		"build:main": "tsc -p tsconfig.build.json",
		"test": "jest"
	},
	"engines": {
		"node": ">=14.18.0"
	},
	"prettier": "@sofie-automation/code-standard-preset/.prettierrc.json",
	"lint-staged": {
		"*.{css,json,md,scss}": [
			"prettier"
		],
		"*.{ts,tsx,js,jsx}": [
			"eslint"
		]
	},
	"files": [
		"/dist",
		"/CHANGELOG.md",
		"/README.md",
		"/LICENSE"
	]
}
