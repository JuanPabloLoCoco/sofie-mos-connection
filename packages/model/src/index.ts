export * from './dataTypes'
export * from './model'
