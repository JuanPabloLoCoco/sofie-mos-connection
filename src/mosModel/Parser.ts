import * as XMLBuilder from 'xmlbuilder'
import {
	IMOSRunningOrder,
	IMOSROStory,
	IMOSItem,
	IMOSObjectPath,
	IMOSRunningOrderBase,
	IMOSObject,
	IMOSROFullStory,
	IMOSRequestObjectList,
	IMOSAck,
} from '../api'
import { IMOSExternalMetaData } from '../dataTypes/mosExternalMetaData'
import { MosString128 } from '../dataTypes/mosString128'
import {
	XMLRunningOrderBase,
	XMLRunningOrder,
	XMLROStories,
	XMLROStory,
	XMLMosItems,
	XMLMosItem,
	XMLObjectPaths,
	XMLMosExternalMetaData,
	XMLMosROAck,
} from './profile2/xmlConversion'
import { XMLROFullStory } from './profile4/xmlConversion'
import { XMLMosIDs } from './profile0/xmlConversion'
import { XMLMosAck, XMLMosObjects, XMLMosObject, XMLMosRequestObjectList } from './profile1/xmlConversion'
import { ROAck } from './profile2/ROAck'
import { AnyXML } from './lib'

/* eslint-disable @typescript-eslint/no-namespace */
export namespace Parser {
	export function xml2ROBase(xml: AnyXML): IMOSRunningOrderBase {
		return XMLRunningOrderBase.fromXML(xml)
	}
	export function xml2RO(xml: AnyXML): IMOSRunningOrder {
		return XMLRunningOrder.fromXML(xml)
	}
	// export function ro2xml (ro: IMOSRunningOrder): XMLBuilder.XMLElement {
	// 	// too implement
	// 	return XMLBuilder.create('ro')
	// }
	export function xml2Stories(xml: Array<any>): Array<IMOSROStory> {
		return XMLROStories.fromXML(xml)
	}
	export function xml2FullStory(xml: AnyXML): IMOSROFullStory {
		return XMLROFullStory.fromXML(xml)
	}
	export function xml2Story(xml: AnyXML): IMOSROStory {
		return XMLROStory.fromXML(xml)
	}
	// export function story2xml (story: IMOSROStory): XMLBuilder.XMLElement {
	// 	return ROStory.toXML(story)
	// }
	export function xml2Items(xml: Array<any>): Array<IMOSItem> {
		return XMLMosItems.fromXML(xml)
	}
	export function xml2Item(xml: AnyXML): IMOSItem {
		return XMLMosItem.fromXML(xml)
	}
	export function xml2ObjPaths(xml: AnyXML): Array<IMOSObjectPath> {
		return XMLObjectPaths.fromXML(xml)
	}
	// export function objPaths2xml (paths: Array<IMOSObjectPath>): XMLBuilder.XMLElement {
	// 	return ObjectPaths.toXML(xmlItem, paths)
	// }
	// export function item2xml (item: IMOSItem): XMLBuilder.XMLElement {
	// 	MOSItem.toXML(xmlItem, item)
	// }
	export function xml2MetaData(xml: AnyXML): Array<IMOSExternalMetaData> {
		return XMLMosExternalMetaData.fromXML(xml)
	}
	export function metaData2xml(md: IMOSExternalMetaData): XMLBuilder.XMLElement {
		// let xmlMD = XMLBuilder.create('mosExternalMetadata')

		// if (md.MosScope) addTextElement(xmlMD, 'mosScope', {}, md.MosScope)
		// addTextElement(xmlMD, 'mosSchema', {}, md.MosSchema)

		// let payload = parser.toXml(md.MosPayload)  // TODO: implement this properly, convert to xml
		// let payload = js2xml({ mosExternalMetadata: md }, { compact: true })
		return XMLBuilder.create({ mosExternalMetadata: md })
		// addTextElement(xmlMD, 'mosPayload', {}, payload)
		// return xmlMD
	}
	export function xml2IDs(xml: AnyXML): Array<MosString128> {
		return XMLMosIDs.fromXML(xml)
	}
	export function xml2ROAck(xml: AnyXML): ROAck {
		return XMLMosROAck.fromXML(xml)
	}
	export function xml2Ack(xml: AnyXML): IMOSAck {
		return XMLMosAck.fromXML(xml)
	}
	export function xml2MosObjs(xml: AnyXML): Array<IMOSObject> {
		return XMLMosObjects.fromXML(xml)
	}
	export function xml2MosObj(xml: AnyXML): IMOSObject {
		return XMLMosObject.fromXML(xml)
	}
	// export function mosObj2xml (obj: IMOSObject): XMLBuilder.XMLElement {
	// 	let xml = XMLBuilder.create('mosObj')
	// 	MOSObject.toXML(xml, obj)
	// 	// Todo: metadata:
	// 	return xml
	// }
	export function xml2ReqObjList(xml: AnyXML): IMOSRequestObjectList {
		return XMLMosRequestObjectList.fromXML(xml)
	}
}
