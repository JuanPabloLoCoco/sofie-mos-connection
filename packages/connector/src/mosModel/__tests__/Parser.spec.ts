import { Parser } from '../Parser'
import { IMOSObjectPathType } from '@mos-connection/model'
import { xml2js } from '../../utils/Utils'

describe('Parser', () => {
	test('xml2ObjPaths: several objPaths', () => {
		const parsed: any = xml2js(`
			<item>
				<itemID>V_700</itemID>
				<itemSlug>AVID MEDIA CENTRAL</itemSlug>
				<objID>3afb7390-7f0c-4d52-ab74-159fad1df7ba</objID>
				<mosID>mam</mosID>
				<mosAbstract></mosAbstract>
				<objPaths>
					<objPath techDescription="VIDEO">http://media.example.com.br:9000/mamfolders/Original/3afb7390-7f0c-4d52-ab74-159fad1df7ba.mp4</objPath>
					<objProxyPath techDescription="JPG">http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_0.jpg</objProxyPath>
					<objProxyPath techDescription="VIDEO">http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_1.mp4</objProxyPath>
				</objPaths>
				<itemChannel></itemChannel>
				<itemEdStart></itemEdStart>
				<itemEdDur>49550</itemEdDur>
				<itemUserTimingDur></itemUserTimingDur>
				<itemTrigger>MANUAL</itemTrigger>
				<macroIn></macroIn>
				<macroOut></macroOut>
			</item>
		`)
		expect(parsed.item).toBeTruthy()
		expect(parsed.item.objPaths).toBeTruthy()
		expect(Parser.xml2ObjPaths(parsed.item.objPaths)).toEqual([
			{
				Type: IMOSObjectPathType.PATH,
				Description: 'VIDEO',
				Target: 'http://media.example.com.br:9000/mamfolders/Original/3afb7390-7f0c-4d52-ab74-159fad1df7ba.mp4',
			},
			{
				Type: IMOSObjectPathType.PROXY_PATH,
				Description: 'JPG',
				Target: 'http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_0.jpg',
			},
			{
				Type: IMOSObjectPathType.PROXY_PATH,
				Description: 'VIDEO',
				Target: 'http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_1.mp4',
			},
		])
		expect(Parser.xml2ObjPaths(parsed.item.objPaths)).toMatchSnapshot()
	})
	test('xml2ObjPaths: only one objPath', () => {
		const parsed: any = xml2js(`
			<item>
				<itemID>V_700</itemID>
				<itemSlug>AVID MEDIA CENTRAL</itemSlug>
				<objID>3afb7390-7f0c-4d52-ab74-159fad1df7ba</objID>
				<mosID>mam</mosID>
				<mosAbstract></mosAbstract>
				<objPaths>
					<objPath techDescription="">D:\\Videos\\1080p\\Wonderful Universe - Part 2 FULL HD.mp4</objPath>
				</objPaths>
				<itemChannel></itemChannel>
				<itemEdStart></itemEdStart>
				<itemEdDur>49550</itemEdDur>
				<itemUserTimingDur></itemUserTimingDur>
				<itemTrigger>MANUAL</itemTrigger>
				<macroIn></macroIn>
				<macroOut></macroOut>
			</item>
		`)
		expect(parsed.item).toBeTruthy()
		expect(parsed.item.objPaths).toBeTruthy()
		expect(Parser.xml2ObjPaths(parsed.item.objPaths)).toEqual([
			{
				Type: IMOSObjectPathType.PATH,
				Description: '',
				Target: 'D:\\Videos\\1080p\\Wonderful Universe - Part 2 FULL HD.mp4',
			},
		])
		expect(Parser.xml2ObjPaths(parsed.item.objPaths)).toMatchSnapshot()
	})
	test('xml2MetaData: handle time formats with various decimal separators', () => {
		const parsed: any = xml2js(`
		<mosExternalMetadata>
		<mosScope>PLAYLIST</mosScope>
		<mosSchema>http://ROENPS01:10505/schema/enps.dtd</mosSchema>
		<mosPayload>
			<Approved>0</Approved>
			<Creator>LINUXENPS</Creator>
			<Estimated>20</Estimated>
			<MediaTime>44.6</MediaTime>
			<MediaTime2>44,6</MediaTime2>
			<ModBy>TEDIAL.ROLXTEDPRDMAM01.DKRO.NRK.MOS</ModBy>
			<ModTime>20190401T161849Z</ModTime>
			<MOSItemDurations>27,56


	</MOSItemDurations>
			<MOSItemEdDurations/>
			<MOSObjSlugs>TEASER-OLD-IRISH-010419-RO
	M: Old Irish Pub kan bli stoppet
	10 Tema|Vanlig (00:00, Auto/OnNext): Pub-kjede hisset på seg politikere
	Story status</MOSObjSlugs>
			<MOSSlugs>TEASER-OLD-IRISH;TEASER-OLD-IRISH-010419-RO-1
	M: Old Irish Pub kan bli stoppet (01-04-19 17:59)
	TEASER-OLD-IRISH;TEASER-OLD-IRISH-010419-RO-4
	SAK 2-17</MOSSlugs>
			<MOSStatus>READY


	</MOSStatus>
			<Owner>LINUXENPS</Owner>
			<SourceMediaTime>0</SourceMediaTime>
			<SourceTextTime>0</SourceTextTime>
			<StoryLogPreview>Den danske pub-kjeda The Old Irish Pub har brutt fleire regelverk allerede før åpninga i Stavanger.  Det faller ikkje i god jord hos politikarane, som nå vil stoppa nyetableringa. </StoryLogPreview>
			<TextTime>12</TextTime>
			<Fylke>Rogaland</Fylke>
			<Innslagstittel>Old Irish Pub kan bli stoppet</Innslagstittel>
			<Kommune>Stavanger</Kommune>
			<mosartTransition>effect 2</mosartTransition>
			<mosartType>STK</mosartType>
			<OpprLand>Norge</OpprLand>
			<ReadTime>12.3</ReadTime>
			<ReadTime2>12,3</ReadTime2>
			<Diary>Dear diary, today I ordered number 1,2 and 4 from the resturant</Diary>
			<ResturantOrders>1,2,4</ResturantOrders>
			<Rettigheter>Grønt</Rettigheter>
			<Rettighetseier>NRK</Rettighetseier>
			<Sted>Stavanger</Sted>
			<Tags>nattklubb, de røde sjøhus, skjenkebevilgning</Tags>
			<AUDIOGAIN/>
			<ENPSItemType>3</ENPSItemType>
		</mosPayload>
	</mosExternalMetadata>
		`)
		expect(parsed).toBeTruthy()
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.MediaTime).toBe(44.6)
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.MediaTime2).toBe(44.6)
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.ReadTime).toBe(12.3)
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.ReadTime2).toBe(12.3)
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.Diary).toBe(
			'Dear diary, today I ordered number 1,2 and 4 from the resturant'
		)
		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)[0].MosPayload.ResturantOrders).toBe('1,2,4')

		expect(Parser.xml2MetaData(parsed.mosExternalMetadata)).toMatchSnapshot()
	})
})
