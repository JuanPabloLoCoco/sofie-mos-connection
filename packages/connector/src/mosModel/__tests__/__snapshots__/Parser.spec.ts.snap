// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser xml2MetaData: handle time formats with various decimal separators 1`] = `
Array [
  Object {
    "MosPayload": Object {
      "AUDIOGAIN": "",
      "Approved": 0,
      "Creator": "LINUXENPS",
      "Diary": "Dear diary, today I ordered number 1,2 and 4 from the resturant",
      "ENPSItemType": 3,
      "Estimated": 20,
      "Fylke": "Rogaland",
      "Innslagstittel": "Old Irish Pub kan bli stoppet",
      "Kommune": "Stavanger",
      "MOSItemDurations": 27.56,
      "MOSItemEdDurations": "",
      "MOSObjSlugs": "TEASER-OLD-IRISH-010419-RO
	M: Old Irish Pub kan bli stoppet
	10 Tema|Vanlig (00:00, Auto/OnNext): Pub-kjede hisset på seg politikere
	Story status",
      "MOSSlugs": "TEASER-OLD-IRISH;TEASER-OLD-IRISH-010419-RO-1
	M: Old Irish Pub kan bli stoppet (01-04-19 17:59)
	TEASER-OLD-IRISH;TEASER-OLD-IRISH-010419-RO-4
	SAK 2-17",
      "MOSStatus": "READY",
      "MediaTime": 44.6,
      "MediaTime2": 44.6,
      "ModBy": "TEDIAL.ROLXTEDPRDMAM01.DKRO.NRK.MOS",
      "ModTime": "20190401T161849Z",
      "OpprLand": "Norge",
      "Owner": "LINUXENPS",
      "ReadTime": 12.3,
      "ReadTime2": 12.3,
      "ResturantOrders": "1,2,4",
      "Rettigheter": "Grønt",
      "Rettighetseier": "NRK",
      "SourceMediaTime": 0,
      "SourceTextTime": 0,
      "Sted": "Stavanger",
      "StoryLogPreview": "Den danske pub-kjeda The Old Irish Pub har brutt fleire regelverk allerede før åpninga i Stavanger.  Det faller ikkje i god jord hos politikarane, som nå vil stoppa nyetableringa.",
      "Tags": "nattklubb, de røde sjøhus, skjenkebevilgning",
      "TextTime": 12,
      "mosartTransition": "effect 2",
      "mosartType": "STK",
      "text": 0,
    },
    "MosSchema": "http://ROENPS01:10505/schema/enps.dtd",
    "MosScope": "PLAYLIST",
  },
]
`;

exports[`Parser xml2ObjPaths: only one objPath 1`] = `
Array [
  Object {
    "Description": "",
    "Target": "D:\\\\Videos\\\\1080p\\\\Wonderful Universe - Part 2 FULL HD.mp4",
    "Type": "PATH",
  },
]
`;

exports[`Parser xml2ObjPaths: several objPaths 1`] = `
Array [
  Object {
    "Description": "VIDEO",
    "Target": "http://media.example.com.br:9000/mamfolders/Original/3afb7390-7f0c-4d52-ab74-159fad1df7ba.mp4",
    "Type": "PATH",
  },
  Object {
    "Description": "JPG",
    "Target": "http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_0.jpg",
    "Type": "PROXY PATH",
  },
  Object {
    "Description": "VIDEO",
    "Target": "http://media.example.com.br:9000/mamfolders/Proxy/3afb7390-7f0c-4d52-ab74-159fad1df7ba_1.mp4",
    "Type": "PROXY PATH",
  },
]
`;
